<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<link rel="stylesheet" href="style.css" type="text/css">
<meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
<link rel="Start" href="index.html">
<link rel="previous" href="Cycledash.html">
<link rel="next" href="Hisat.html">
<link rel="Up" href="index.html">
<link title="Index of types" rel=Appendix href="index_types.html">
<link title="Index of values" rel=Appendix href="index_values.html">
<link title="Index of modules" rel=Appendix href="index_modules.html">
<link title="Bedtools" rel="Chapter" href="Bedtools.html">
<link title="Bwa" rel="Chapter" href="Bwa.html">
<link title="Common" rel="Chapter" href="Common.html">
<link title="Common_pipelines" rel="Chapter" href="Common_pipelines.html">
<link title="Cufflinks" rel="Chapter" href="Cufflinks.html">
<link title="Cycledash" rel="Chapter" href="Cycledash.html">
<link title="Gatk" rel="Chapter" href="Gatk.html">
<link title="Hisat" rel="Chapter" href="Hisat.html">
<link title="Mosaik" rel="Chapter" href="Mosaik.html">
<link title="Muse" rel="Chapter" href="Muse.html">
<link title="Mutect" rel="Chapter" href="Mutect.html">
<link title="Picard" rel="Chapter" href="Picard.html">
<link title="Pipeline" rel="Chapter" href="Pipeline.html">
<link title="Reference_genome" rel="Chapter" href="Reference_genome.html">
<link title="Region" rel="Chapter" href="Region.html">
<link title="Run_environment" rel="Chapter" href="Run_environment.html">
<link title="Samtools" rel="Chapter" href="Samtools.html">
<link title="Somaticsniper" rel="Chapter" href="Somaticsniper.html">
<link title="Star" rel="Chapter" href="Star.html">
<link title="Strelka" rel="Chapter" href="Strelka.html">
<link title="Stringtie" rel="Chapter" href="Stringtie.html">
<link title="Varscan" rel="Chapter" href="Varscan.html">
<link title="Vcftools" rel="Chapter" href="Vcftools.html">
<link title="Virmid" rel="Chapter" href="Virmid.html">
<link title="Workflow_utilities" rel="Chapter" href="Workflow_utilities.html"><title>Gatk</title>
</head>
<body>
<div class="navbar"><a class="pre" href="Cycledash.html" title="Cycledash">Previous</a>
&nbsp;<a class="up" href="index.html" title="Index">Up</a>
&nbsp;<a class="post" href="Hisat.html" title="Hisat">Next</a>
</div>
<h1>Module <a href="type_Gatk.html">Gatk</a> (<a href="code_Gatk.html">.ml</a>)</h1>

<pre><span class="keyword">module</span> Gatk: <code class="code"><span class="keyword">sig</span></code> <a href="Gatk.html">..</a> <code class="code"><span class="keyword">end</span></code></pre><hr width="100%">

<pre><span id="VALindel_realigner"><span class="keyword">val</span> <a href="code_VALGatk.indel_realigner.html">indel_realigner</a></span> : <code class="type">?compress:bool -><br>       reference_build:<a href="Reference_genome.html#TYPEspecification">Reference_genome.specification</a> -><br>       processors:int -><br>       run_with:<a href="Run_environment.Machine.html#TYPEt">Run_environment.Machine.t</a> -><br>       <a href="Common.KEDSL.html#TYPEbam_file">Common.KEDSL.bam_file</a> <a href="Common.KEDSL.html#TYPEworkflow_node">Common.KEDSL.workflow_node</a> -><br>       output_bam:string -> <a href="Common.KEDSL.html#TYPEbam_file">Common.KEDSL.bam_file</a> <a href="Common.KEDSL.html#TYPEworkflow_node">Common.KEDSL.workflow_node</a></code></pre>
<pre><span id="VALcall_gatk"><span class="keyword">val</span> <a href="code_VALGatk.call_gatk.html">call_gatk</a></span> : <code class="type">analysis:Common.String.t -><br>       ?region:[< `Chromosome of string<br>                | `Chromosome_interval of string * int * int<br>                | `Full<br>                > `Full ] -><br>       string list -> Common.KEDSL.Program.t</code></pre>
<pre><span id="VALbase_quality_score_recalibrator"><span class="keyword">val</span> <a href="code_VALGatk.base_quality_score_recalibrator.html">base_quality_score_recalibrator</a></span> : <code class="type">run_with:<a href="Run_environment.Machine.html#TYPEt">Run_environment.Machine.t</a> -><br>       processors:int -><br>       reference_build:<a href="Reference_genome.html#TYPEspecification">Reference_genome.specification</a> -><br>       input_bam:<a href="Common.KEDSL.html#TYPEbam_file">Common.KEDSL.bam_file</a> <a href="Common.KEDSL.html#TYPEworkflow_node">Common.KEDSL.workflow_node</a> -><br>       output_bam:string -> <a href="Common.KEDSL.html#TYPEbam_file">Common.KEDSL.bam_file</a> <a href="Common.KEDSL.html#TYPEworkflow_node">Common.KEDSL.workflow_node</a></code></pre>
<pre><span id="VALhaplotype_caller"><span class="keyword">val</span> <a href="code_VALGatk.haplotype_caller.html">haplotype_caller</a></span> : <code class="type">?more_edges:<a href="Common.KEDSL.html#TYPEworkflow_edge">Common.KEDSL.workflow_edge</a> list -><br>       run_with:<a href="Run_environment.Machine.html#TYPEt">Run_environment.Machine.t</a> -><br>       reference_build:<a href="Reference_genome.html#TYPEspecification">Reference_genome.specification</a> -><br>       input_bam:<a href="Common.KEDSL.html#TYPEbam_file">Common.KEDSL.bam_file</a> <a href="Common.KEDSL.html#TYPEworkflow_node">Common.KEDSL.workflow_node</a> -><br>       result_prefix:string -><br>       [< `Map_reduce<br>        | `Region of<br>            [< `Chromosome of string<br>             | `Chromosome_interval of string * int * int<br>             | `Full<br>             > `Full ] ] -><br>       <a href="Common.KEDSL.html#TYPEsingle_file">Common.KEDSL.single_file</a> <a href="Common.KEDSL.html#TYPEworkflow_node">Common.KEDSL.workflow_node</a></code></pre></body></html>