<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<link rel="stylesheet" href="style.css" type="text/css">
<meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
<link rel="Start" href="index.html">
<link rel="previous" href="Biokepi_bfx_tools.Picard.html">
<link rel="next" href="Biokepi_bfx_tools.Seq2HLA.html">
<link rel="Up" href="Biokepi_bfx_tools.html">
<link title="Index of types" rel=Appendix href="index_types.html">
<link title="Index of extensions" rel=Appendix href="index_extensions.html">
<link title="Index of values" rel=Appendix href="index_values.html">
<link title="Index of modules" rel=Appendix href="index_modules.html">
<link title="Index of module types" rel=Appendix href="index_module_types.html">
<link title="Biokepi_run_environment" rel="Chapter" href="Biokepi_run_environment.html">
<link title="Biokepi_environment_setup" rel="Chapter" href="Biokepi_environment_setup.html">
<link title="Biokepi_bfx_tools" rel="Chapter" href="Biokepi_bfx_tools.html">
<link title="Biokepi_pipeline_edsl" rel="Chapter" href="Biokepi_pipeline_edsl.html">
<link title="All_downloads" rel="Chapter" href="All_downloads.html">
<link title="Main" rel="Chapter" href="Main.html">
<link title="Ttfi_pipeline" rel="Chapter" href="Ttfi_pipeline.html">
<link title="Biokepi" rel="Chapter" href="Biokepi.html"><title>Biokepi API : Biokepi_bfx_tools.Samtools</title>
</head>
<body>
<div class="navbar"><a class="pre" href="Biokepi_bfx_tools.Picard.html" title="Biokepi_bfx_tools.Picard">Previous</a>
&nbsp;<a class="up" href="Biokepi_bfx_tools.html" title="Biokepi_bfx_tools">Up</a>
&nbsp;<a class="post" href="Biokepi_bfx_tools.Seq2HLA.html" title="Biokepi_bfx_tools.Seq2HLA">Next</a>
</div>
<h1>Module <a href="type_Biokepi_bfx_tools.Samtools.html">Biokepi_bfx_tools.Samtools</a> (<a href="code_Biokepi_bfx_tools.Samtools.html">.ml</a>)</h1>

<pre><span class="keyword">module</span> Samtools: <code class="code"><span class="keyword">sig</span></code> <a href="Biokepi_bfx_tools.Samtools.html">..</a> <code class="code"><span class="keyword">end</span></code></pre><hr width="100%">

<pre><span class="keyword">module</span> <a href="Biokepi_bfx_tools.Samtools.Remove.html">Remove</a>: <code class="type"><a href="Biokepi_run_environment.Workflow_utilities.Remove.html">Biokepi_run_environment.Workflow_utilities.Remove</a></code></pre>
<pre><span id="VALsam_to_bam"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.sam_to_bam.html">sam_to_bam</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       reference_build:string -><br>       < is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; .. ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       < host : Biokepi_run_environment.Common.KEDSL.Host.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; reference_build : string;<br>         sorting : [ `Coordinate | `Read_name ] option ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALfaidx"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.faidx.html">faidx</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       < is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; .. ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       < exists : Ketrew_pure.Target.Condition.t;<br>         is_bigger_than : int -> Ketrew_pure.Target.Condition.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALflagstat"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.flagstat.html">flagstat</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       < is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; .. ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       < exists : Ketrew_pure.Target.Condition.t;<br>         is_bigger_than : int -> Ketrew_pure.Target.Condition.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALbgzip"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.bgzip.html">bgzip</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       < is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; .. ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       string -><br>       < exists : Ketrew_pure.Target.Condition.t;<br>         is_bigger_than : int -> Ketrew_pure.Target.Condition.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALtabix"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.tabix.html">tabix</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       tabular_format:[< `Bed | `Gff | `Psltab | `Sam | `Vcf ] -><br>       < is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; .. ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       < exists : Ketrew_pure.Target.Condition.t;<br>         is_bigger_than : int -> Ketrew_pure.Target.Condition.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALdo_on_bam"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.do_on_bam.html">do_on_bam</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       ?more_depends_on:Biokepi_run_environment.Common.KEDSL.workflow_edge list -><br>       name:string -><br>       ?more_requirements:<a href="Biokepi_run_environment.Machine.Make_fun.Requirement.html#TYPEt">Biokepi_run_environment.Machine.Make_fun.Requirement.t</a><br>                          list -><br>       < is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : 'a; .. ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       product:(< is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>                  path : string; .. ><br>                as 'b)<br>               Biokepi_run_environment.Common.KEDSL.product -><br>       make_command:('a -> string -> string list) -><br>       'b Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALsort_bam_no_check"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.sort_bam_no_check.html">sort_bam_no_check</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       by:[ `Coordinate | `Read_name ] -><br>       < host : Biokepi_run_environment.Common.KEDSL.Host.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; reference_build : string;<br>         sorting : [ `Coordinate | `Read_name ] option ><br>       Biokepi_run_environment.Common.KEDSL.product<br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       < host : Biokepi_run_environment.Common.KEDSL.Host.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; reference_build : string;<br>         sorting : [ `Coordinate | `Read_name ] option ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALsort_bam_if_necessary"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.sort_bam_if_necessary.html">sort_bam_if_necessary</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       by:[ `Coordinate | `Read_name ] -><br>       <a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_file">Biokepi_run_environment.Common.KEDSL.bam_file</a><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       <a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_file">Biokepi_run_environment.Common.KEDSL.bam_file</a><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre><div class="info ">
Uses <code class="code"><span class="string">"samtools sort"</span></code> if the <code class="code">input_bam</code> is not tagged as
    “sorted as the <code class="code">~by</code> argument.”
    If it is indeed already sorted the function returns the <code class="code">input_bam</code> node as
    is.<br>
</div>

<pre><span id="VALindex_to_bai"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.index_to_bai.html">index_to_bai</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       ?check_sorted:bool -><br>       < is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; sorting : [> `Read_name ] option; .. ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       < exists : Ketrew_pure.Target.Condition.t;<br>         is_bigger_than : int -> Ketrew_pure.Target.Condition.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALmpileup"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.mpileup.html">mpileup</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       ?adjust_mapq:int -><br>       region:[< `Chromosome of string<br>               | `Chromosome_interval of string * int * int<br>               | `Full ] -><br>       <a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_file">Biokepi_run_environment.Common.KEDSL.bam_file</a><br>       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       < exists : Ketrew_pure.Target.Condition.t;<br>         is_bigger_than : int -> Ketrew_pure.Target.Condition.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALmerge_bams"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Samtools.merge_bams.html">merge_bams</a></span> : <code class="type">run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       ?delete_input_on_success:bool -><br>       ?attach_rg_tag:bool -><br>       ?uncompressed_bam_output:bool -><br>       ?compress_level_one:bool -><br>       ?combine_rg_headers:bool -><br>       ?combine_pg_headers:bool -><br>       <a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_file">Biokepi_run_environment.Common.KEDSL.bam_file</a><br>       Biokepi_run_environment.Common.KEDSL.workflow_node list -><br>       string -><br>       < host : Biokepi_run_environment.Common.KEDSL.Host.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; reference_build : string;<br>         sorting : [ `Coordinate | `Read_name ] option ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre><div class="info ">
Merge a list of BAM files.
<p>

   The BAMs will be sorted, by coordinate, if they are not already.
<p>

   <code class="code"><span class="string">"samtools merge"</span></code> fails if the list of bams has only one (or zero) bams, so
   this functions raises an exception if <code class="code">input_bam_list</code> has lenth <code class="code">&lt;</code> 2.
<p>

   By default, duplicate @PG and @RG IDs will be automatically uniquified with a
   random suffix.
<p>
<ul>
<li><code class="code">?delete_input_on_success</code>: Delete the list of input BAMs when this node
   succeeds. (default: true)</li>
<li><code class="code">?attach_rg_tag</code>: Attach an RG tag to each alignment. The tag value is
   inferred from file names.(default: false)</li>
<li><code class="code">?uncompressed_bam_output</code>: Uncompressed BAM output. (default: false)</li>
<li><code class="code">?compress_level_one</code>: Use zlib compression level 1 to compress the output.
   (default: false)</li>
<li><code class="code">?combine_rg_headers</code>: When several input files contain @RG headers with the
   same ID, emit only one the first of them. (default: false)</li>
<li><code class="code">?combine_pg_headers</code>: When several input files contain @PG headers with the
   same ID, emit only one the first of them. (default: false)</li>
</ul>
<br>
</div>
</body></html>