<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<link rel="stylesheet" href="style.css" type="text/css">
<meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
<link rel="Start" href="index.html">
<link rel="previous" href="Biokepi_bfx_tools.Cycledash.html">
<link rel="next" href="Biokepi_bfx_tools.Hisat.html">
<link rel="Up" href="Biokepi_bfx_tools.html">
<link title="Index of types" rel=Appendix href="index_types.html">
<link title="Index of extensions" rel=Appendix href="index_extensions.html">
<link title="Index of values" rel=Appendix href="index_values.html">
<link title="Index of modules" rel=Appendix href="index_modules.html">
<link title="Index of module types" rel=Appendix href="index_module_types.html">
<link title="Biokepi_run_environment" rel="Chapter" href="Biokepi_run_environment.html">
<link title="Biokepi_environment_setup" rel="Chapter" href="Biokepi_environment_setup.html">
<link title="Biokepi_bfx_tools" rel="Chapter" href="Biokepi_bfx_tools.html">
<link title="Biokepi_pipeline_edsl" rel="Chapter" href="Biokepi_pipeline_edsl.html">
<link title="All_downloads" rel="Chapter" href="All_downloads.html">
<link title="Main" rel="Chapter" href="Main.html">
<link title="Ttfi_pipeline" rel="Chapter" href="Ttfi_pipeline.html">
<link title="Biokepi" rel="Chapter" href="Biokepi.html"><title>Biokepi API : Biokepi_bfx_tools.Gatk</title>
</head>
<body>
<div class="navbar"><a class="pre" href="Biokepi_bfx_tools.Cycledash.html" title="Biokepi_bfx_tools.Cycledash">Previous</a>
&nbsp;<a class="up" href="Biokepi_bfx_tools.html" title="Biokepi_bfx_tools">Up</a>
&nbsp;<a class="post" href="Biokepi_bfx_tools.Hisat.html" title="Biokepi_bfx_tools.Hisat">Next</a>
</div>
<h1>Module <a href="type_Biokepi_bfx_tools.Gatk.html">Biokepi_bfx_tools.Gatk</a> (<a href="code_Biokepi_bfx_tools.Gatk.html">.ml</a>)</h1>

<pre><span class="keyword">module</span> Gatk: <code class="code"><span class="keyword">sig</span></code> <a href="Biokepi_bfx_tools.Gatk.html">..</a> <code class="code"><span class="keyword">end</span></code></pre><hr width="100%">

<pre><span class="keyword">module</span> <a href="Biokepi_bfx_tools.Gatk.Remove.html">Remove</a>: <code class="type"><a href="Biokepi_run_environment.Workflow_utilities.Remove.html">Biokepi_run_environment.Workflow_utilities.Remove</a></code></pre>
<pre><span class="keyword">module</span> <a href="Biokepi_bfx_tools.Gatk.Configuration.html">Configuration</a>: <code class="code"><span class="keyword">sig</span></code> <a href="Biokepi_bfx_tools.Gatk.Configuration.html">..</a> <code class="code"><span class="keyword">end</span></code></pre>
<pre><span id="VALindel_realigner_output_filename_tag"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Gatk.indel_realigner_output_filename_tag.html">indel_realigner_output_filename_tag</a></span> : <code class="type">configuration:Configuration.Indel_realigner.t *<br>                     Configuration.Realigner_target_creator.t -><br>       ?region:[< `Chromosome of string<br>                | `Chromosome_interval of string * int * int<br>                | `Full ] -><br>       < product : < path : string; .. >; .. > list -> string</code></pre>
<pre><span id="VALindel_realigner"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Gatk.indel_realigner.html">indel_realigner</a></span> : <code class="type">'a.<br>         ?compress:bool -><br>         ?on_region:<a href="Biokepi_run_environment.Region.html#TYPEt">Biokepi_run_environment.Region.t</a> -><br>         configuration:Configuration.Indel_realigner.t *<br>                       Configuration.Realigner_target_creator.t -><br>         run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>         ?run_directory:string -><br>         'a <a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_or_bams">Biokepi_run_environment.Common.KEDSL.bam_or_bams</a> -> 'a</code></pre>
<pre><span id="VALindel_realigner_map_reduce"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Gatk.indel_realigner_map_reduce.html">indel_realigner_map_reduce</a></span> : <code class="type">'a.<br>         ?compress:bool -><br>         configuration:Configuration.Indel_realigner.t *<br>                       Configuration.Realigner_target_creator.t -><br>         run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>         ?run_directory:string -><br>         'a <a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_or_bams">Biokepi_run_environment.Common.KEDSL.bam_or_bams</a> -> 'a</code></pre>
<pre><span id="VALcall_gatk"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Gatk.call_gatk.html">call_gatk</a></span> : <code class="type">analysis:string -><br>       ?region:[< `Chromosome of string<br>                | `Chromosome_interval of string * int * int<br>                | `Full<br>                > `Full ] -><br>       string list -> Biokepi_run_environment.Common.KEDSL.Program.t</code></pre>
<pre><span id="VALbase_quality_score_recalibrator"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Gatk.base_quality_score_recalibrator.html">base_quality_score_recalibrator</a></span> : <code class="type">configuration:Configuration.Bqsr.t *<br>                     Configuration.Print_reads.t -><br>       run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       input_bam:<a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_file">Biokepi_run_environment.Common.KEDSL.bam_file</a><br>                 Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       output_bam:string -><br>       < host : Biokepi_run_environment.Common.KEDSL.Host.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string; reference_build : string;<br>         sorting : [ `Coordinate | `Read_name ] option ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALhaplotype_caller"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Gatk.haplotype_caller.html">haplotype_caller</a></span> : <code class="type">?more_edges:Biokepi_run_environment.Common.KEDSL.workflow_edge list -><br>       run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       input_bam:<a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_file">Biokepi_run_environment.Common.KEDSL.bam_file</a><br>                 Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       result_prefix:string -><br>       [< `Map_reduce<br>        | `Region of<br>            [< `Chromosome of string<br>             | `Chromosome_interval of string * int * int<br>             | `Full<br>             > `Full ] ] -><br>       < exists : Ketrew_pure.Target.Condition.t;<br>         is_bigger_than : int -> Ketrew_pure.Target.Condition.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre>
<pre><span id="VALmutect2"><span class="keyword">val</span> <a href="code_VALBiokepi_bfx_tools.Gatk.mutect2.html">mutect2</a></span> : <code class="type">?more_edges:Biokepi_run_environment.Common.KEDSL.workflow_edge list -><br>       configuration:<a href="Biokepi_bfx_tools.Gatk.Configuration.Mutect2.html#TYPEt">Configuration.Mutect2.t</a> -><br>       run_with:<a href="Biokepi_run_environment.Machine.html#TYPEt">Biokepi_run_environment.Machine.t</a> -><br>       input_normal_bam:<a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_file">Biokepi_run_environment.Common.KEDSL.bam_file</a><br>                        Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       input_tumor_bam:<a href="Biokepi_run_environment.Common.KEDSL.html#TYPEbam_file">Biokepi_run_environment.Common.KEDSL.bam_file</a><br>                       Biokepi_run_environment.Common.KEDSL.workflow_node -><br>       result_prefix:string -><br>       [< `Map_reduce<br>        | `Region of<br>            [< `Chromosome of string<br>             | `Chromosome_interval of string * int * int<br>             | `Full<br>             > `Full ] ] -><br>       < exists : Ketrew_pure.Target.Condition.t;<br>         is_bigger_than : int -> Ketrew_pure.Target.Condition.t;<br>         is_done : Biokepi_run_environment.Common.KEDSL.Condition.t option;<br>         path : string ><br>       Biokepi_run_environment.Common.KEDSL.workflow_node</code></pre><div class="info ">
Call somatic variants with Mutect2.
<p>

    Mutect2 comes within the GATK (as opposed to <a href="Biokepi_bfx_tools.Mutect.html"><code class="code"><span class="constructor">Biokepi_bfx_tools</span>.<span class="constructor">Mutect</span></code></a>).
<p>

    Cf. also
    https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_cancer_m2_MuTect2.php<br>
</div>
</body></html>